pydocstringformatter/run.py

```diff
@@ -98,26 +84,148 @@ def format_file(self, filename: Path) -> bool:
                         file=sys.stderr,
                     )
                 with open(filename, "w", encoding="utf-8", newline=newlines) as file:
-                    file.write(tokenize.untokenize(changed_tokens))
+                    file.write(tokenize.untokenize(formatted_tokens))
                     _utils.print_to_console(
                         f"Formatted {filename_str} 📖\n", self.config.quiet
                     )
             else:
                 sys.stdout.write(
                     _utils.generate_diff(
                         tokenize.untokenize(tokens),
-                        tokenize.untokenize(changed_tokens),
+                        tokenize.untokenize(formatted_tokens),
                         filename_str,
                     )
                 )
 
         return is_changed
 
-    def format_files(self, filepaths: list[Path]) -> bool:
-        """Format a list of files."""
+    def format_file_tokens(
+        self, tokens: list[tokenize.TokenInfo], filename: Path
+    ) -> tuple[list[tokenize.TokenInfo], bool]:
+        """Format a list of tokens.
+
+        tokens: List of tokens to format.
+        filename: Name of the file the tokens are from.
+
+        Returns:
+            A tuple containing [1] the formatted tokens in a list
+            and a boolean indicating if the tokens were changed.
+
+        Raises:
+            RuntimeError:
+                If the formatters are not able to get to a stable
+                result after 3 iterations.
+
+                It reports what formatters are still modifying the tokens.
+        """
+        formatted_tokens: list[tokenize.TokenInfo] = []
         is_changed = False
 
-        for file in filepaths:
-            is_changed = self.format_file(file) or is_changed
+        for index, tokeninfo in enumerate(tokens):
+            new_tokeninfo = tokeninfo
 
-        return is_changed
+            if _utils.is_docstring(new_tokeninfo, tokens[index - 1]):
+                new_tokeninfo, changers, stable = self.apply_formatters_twice(
+                    new_tokeninfo
+                )
+                is_changed = is_changed or bool(changers)
+
+                # Try again to get a stable result, skipping the formatters that
+                # didn't change the tokens in former passes
+                new_tokeninfo, changers, stable = self.apply_formatters_twice(
+                    new_tokeninfo, changers
+                )
+
+                if not stable:
+                    if len(changers) > 1:
+                        msg = "Conflicting formatters: " + ", ".join(changers)
+                    else:
+                        msg = "Formatter: " + changers[0]
```

Comment:
I like that we also give a the name of the formatter that is "broken". Good idea!